Hermes/appscale/hermes/stats/pubsub_base.py

```diff
@@ -0,0 +1,147 @@
+""" Base classes for building stats publishers and subscribers """
+import logging
+import sys
+
+from tornado.ioloop import PeriodicCallback, IOLoop
+
+
+class SubscriberIsAlreadyRegistered(Exception):
+  """ Raised when you register single subscriber twice on one publisher """
+  pass
+
+
+class SubscriberIsNotRegistered(Exception):
+  """ Raised when you unregister subscriber which is not registered yet """
+  pass
+
+
+class StatsSource(object):
+  """
+  Base class for producers of any kind of stats.
+  Subclasses should implement method get_current()
+  """
+  @property
+  def stats_name(self):
+    return self.__class__.__name__
+
+  def __repr__(self):
+    return self.stats_name
+
+  def get_current(self):
+    """ Returns: current value of specific kind of stats """
+    raise NotImplementedError()
+
+
+class AsyncStatsSource(object):
+  """
+  Base class for asynchronous producers of any kind of stats.
+  Subclasses should implement get_current_async()
+  """
+  @property
+  def stats_name(self):
+    return self.__class__.__name__
+
+  def __repr__(self):
+    return self.stats_name
+
+  def get_current_async(self):
+    """ Returns: Future wrapper for current value of specific kind of stats """
+    raise NotImplementedError()
+
+
+class StatsSubscriber(object):
+  """
+  Base class for stats consumers.
+  Subclasses should implement method receive(stats)
+  """
+  @property
+  def subscriber_name(self):
+      return self.__class__.__name__
+
+  def __repr__(self):
+    return self.subscriber_name
+
+  def receive(self, stats):
+    """ Handlers produced stats entity
+    Args:
+      stats: an object containing stats of some kind (node, processes, ...)
+    """
+    raise NotImplementedError()
+
+
+class StatsPublisher(object):
```

Comment:
During our last team discussion about it, the profile log was not going to be permanent. It was just for demonstration purposes. Conceptually, the portal is the same as any other client that is polling Hermes. The mechanism is different because of potential firewall issues, so it might make sense for the http client to be a different actor or process.

I've just never heard of a local cache "subscribing" to anything. With this terminology, clients that are asking for stats are not consumers, but the service's cache is. If you feel strongly about it, I will approve. It just seems counter-intuitive to me.